File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
     int n=s.size();
     int maxlen=0;
        unordered_map<char, int> ans;
        int l = 0;

        for (int r = 0; r < n; r++) {
            if (ans.find(s[r]) != ans.end() && ans[s[r]] >= l) {
                l = ans[s[r]] + 1;
            }

            ans[s[r]] = r;
            int len = r - l + 1;
            maxlen = max(len, maxlen);
        }

        return maxlen;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n=len(s)
        l=0
        r=0
        if n==0:
            return 0
        dict1={}
        maxs=1
        while l<n and r<n:
            ch=s[r]
            ch1=s[l]
            if dict1.get(ch,0)==0:
                dict1[ch]=dict1.get(ch,0)+1
                a=r-l+1
                maxs=max(maxs,a)
                r+=1
            else :
                dict1[ch1]=dict1.get(ch1,0)-1
                l+=1

            
        return maxs
                

            
        